src/main/java/pl/damianszczepanik/jenkins/buildhistorymanager/model/actions/DeleteArtifactsMatchingPatternsAction.java

```diff
@@ -0,0 +1,168 @@
+    }
+
+    @DataBoundSetter
+    public void setIncludePatterns(String includePatterns) {
+        this.includePatterns = includePatterns;
+    }
+
+    public String getExcludePatterns() {
+        return excludePatterns;
+    }
+
+    @DataBoundSetter
+    public void setExcludePatterns(String excludePatterns) {
+        this.excludePatterns = excludePatterns;
+    }
+
+    static boolean isDirectoryEmpty(Path path) throws IOException {
+        try (DirectoryStream<Path> directoryStream = Files.newDirectoryStream(path)) {
+            return !directoryStream.iterator().hasNext();
+        }
+    }
+
+    // if 'file' is on a different node, this FileCallable will be transferred to that node and executed there.
+    public static final class Delete implements FileCallable<Void> {
+        private static final long serialVersionUID = -7712623306942117400L;
+        private final File archiveRootFile;
+
+        public Delete(File archiveRootFile) {
+            this.archiveRootFile = archiveRootFile;
+        }
+
+        @Override 
+        public Void invoke(File file, VirtualChannel channel) throws IOException {
+            Set<File> directories = new HashSet<>();
+            deleteFileOrLogError(file, directories);
+            deleteEmptyDirectoriesAndParents(directories);
+            return null;
+        }
+
+        void deleteEmptyDirectoriesAndParents(Set<File> directories) throws IOException {
+            for (File dir : directories) {
+                if (shouldDeleteDirectory(dir)) {
+                    Util.deleteFile(dir);
+                    deleteParentDirectories(dir.getParentFile());
+                }
+            }
+        }
+
+        boolean shouldDeleteDirectory(File dir) throws IOException {
+            if (!isValidDirectory(dir)) {
+                return false;
+            }
+
+            File parent = dir.getParentFile();
+            return hasValidParent(parent);
+        }
+
+        boolean hasValidParent(File parent) {
+            return this.archiveRootFile.equals(parent);
+        }
+
+        boolean isValidDirectory(File dir) throws IOException {
+            return dir != null && dir.isDirectory() && isDirectoryEmpty(dir.toPath());
+        }
+
+        void deleteParentDirectories(File directory) {
+            File parent = directory;
+
+            while (shouldDelete(parent)) {
+                deleteDirectory(parent);
+                parent = parent.getParentFile();
+            }
+        }
+
+        boolean shouldDelete(File directory) {
+            return directory != null && !directory.equals(this.archiveRootFile);
+        }
+
+        void deleteDirectory(File directory) {
+            boolean deleteSuccess = directory.delete();
+            if(!deleteSuccess) {
+                LOG.log(Level.FINE, "Deletion of directory failed: " + directory.getAbsolutePath());
+            }
+        }
+
+        void deleteFileOrLogError(File file, Set<File> directories) throws IOException {
```

Comment:
there is to many code lines about deleting directory - why you are going to implement this on your own instead of using library?
I have found some example like https://www.baeldung.com/java-delete-directory that makes deletion in one or two lines

This is definitely the approach which I would like to go for.

Reply:
To be fair, the code started simpler but one of the code analysis rules kept nagging about methods being "too large", then code coverage rules also encouraged extracting small methods for more focused testing.  While there's _some_ simplification possible, there are some constraints in play that make that more difficult.